### PR TITLE
Implement websocket support in Dropshot as an Extractor and #[channel] macro

### PR DESCRIPTION
**Pull request by lifning**
This allows endpoints meant for handling continuous websocket streams to be easily constructed by providing an async handler to an extractor parameter in the endpoint function definition.

This change also refactors the previous `paginated` bool to instead be a dedicated type describing whether an endpoint uses pagination or websockets Dropshot-specific extensions, aspirationally leaving room for describing any more extension modes we may add atop the basic OpenAPI.

### CODE REVIEW

File: dropshot/src/api_description.rs
Comment (david-crespo) [inline dropshot/src/api_description.rs:81]: So this means a given endpoint can only be in one mode at a time, right? That makes sense for pagination and websockets, since they're mutually exclusive. I have no idea what other kinds of extensions we might dream up, but is it possible we could want to use more than one at a time?

File: dropshot/examples/websocket.rs
Comment (david-crespo) [inline dropshot/examples/websocket.rs:74]: amazing API for this

File: dropshot/src/api_description.rs
Comment (lifning) [reply dropshot/src/api_description.rs:81]: yeah - i suppose if such a thing comes to pass, we could change to be something more sophisticated than a bare enum, but i didn't want to try to anticipate unknown unknowns yet, just represent that pagination and websockets are mutually exclusive extensions

File: dropshot/Cargo.toml
Comment (luqmana) [inline dropshot/Cargo.toml:36]: A lil unfortunate we have to pull tungstenite + all its deps just for one function (`derive_accept_key`). It's small enough and not really subject to change that I almost want to say we should just pull it in rather than the whole crate.

**Review by luqmana — COMMENTED** [review]
I left a couple small comments but otherwise looks good, thanks @lifning!

File: dropshot/src/websocket.rs
Comment (lifning) [reply dropshot/src/websocket.rs:196]: Sure! I think I'm more inclined to just make a struct for it to make it a little more ergonomic/readable/explicit than a plain tuple and having people remember to `.0` and `.1`

File: dropshot/src/api_description.rs
Comment (lifning) [reply dropshot/src/api_description.rs:1601]: Ahh, I wrote this before the rust-toolchain.toml bump to 1.62 and didn't think to revisit it!

File: dropshot/src/websocket.rs
Comment (luqmana) [inline dropshot/src/websocket.rs:174]: These should be already be inherited from `rqctx` I thought. +1 on the upgrade one tho, useful to differentiate which logs are from what component.

Comment (lifning) [reply dropshot/src/websocket.rs:174]: gotcha -- also fwiw in case it's not clear, this one is just used by the `error!` calls in the wrapper around the user-provided future in `WebsocketUpgrade::handle`; if the user needs to log in their future, they can obtain it from rqctx the same way they do in any other endpoint & differentiate it however they wish

Comment (luqmana) [reply dropshot/src/websocket.rs:174]: ah, you're right. that makes sense!

**Review by luqmana — APPROVED** [review]

File: dropshot/src/websocket.rs
Comment (davepacheco) [inline dropshot/src/websocket.rs:38]: Why's that?  (why not implicitly call `do_not_upgrade()` on drop, say?)

It looks like it's a non-fatal error to handle the upgrade more than once (we produce an error) but it's fatal to not handle it at all.  I don't have a strong idea here but I just wonder if there's a way to make it harder to accidentally panic in production.  (One thought I had was that `WebsocketUpgrade` could be parametrized by an impl of some trait that would allow us to immediately and automatically invoke it exactly once for you...but then I'm not sure what the body of the function would do.  Also, I'm not sure if it's desirable to be able to do stuff or look at the request before the upgrade -- presumably it is (e.g., authentication).  Then I thought this could be a different kind of endpoint...but then I worry I'm overthinking it.)

File: dropshot/src/websocket.rs
Comment (davepacheco) [inline dropshot/src/websocket.rs:43]: Did you mean to put some docs here and on `WebsocketUpgradeResponse`?

**Review by davepacheco — COMMENTED** [review]
Cool! I had a few questions.

Would you mind updating the changelog to describe the new feature?

File: dropshot/src/websocket.rs
Comment (lifning) [reply dropshot/src/websocket.rs:38]: for the interface to be more discoverable (because `.handle()` *has* to be called for any of this to work; this way nobody's wondering why their connection isn't getting upgraded, even if they weren't originally clear on the docs) and less prone to incorrect use (e.g. accidental inclusion of a WebsocketUpgrade in a param list due to copy/paste errors, causing us to eventually generate an confusing/incorrect client from the resulting schema)

if we're very concerned about panics making it into production, we can make it be an error log in Drop instead, but that's less likely to be caught at development time -- maybe panic if debug assertions are on, else log?

File: dropshot/src/websocket.rs
Comment (lifning) [reply dropshot/src/websocket.rs:43]: :woman_facepalming: sure did! thanks

File: dropshot/src/handler.rs
Comment (lifning) [reply dropshot/src/handler.rs:235]: it means the two tuple elements were indicators of different non-None extension modes (i.e. Paginated and Websocket at this point), in either order, which we currently take to be mutually exclusive things

**Review by ahl — COMMENTED** [review]
Are other extractors permitted alongside `WebsocketUpgrade`? Or must it be the lone extractor?

What does this look like in the OpenAPI output? The OpenAPI spec folks have pretty much declared Websockets as "not HTTP APIs" and therefore out of scope.

I vaguely had intended to use [AsyncAPI](https://www.asyncapi.com/) to describe websocket interfaces, both the channels and the messages. AsyncAPI and OpenAPI are at least sort of compatible in that I think we could produce a single JSON file with both descriptions in there.

Does it make sense to use the same `endpoint` macro with the constraints that the method must be a `GET`, the `WebsocketUpgrade` must be present, and the return type is `Result<Response<Body>, HttpError>`? Or might we want some other interface such as `#[channel { path = ... } ]` or `#[endpoint { method = WEBSOCKET, ...`?

File: dropshot/src/lib.rs
Comment (ahl) [inline dropshot/src/lib.rs:669]: Why is this pub?

File: dropshot/src/websocket.rs
Comment (ahl) [inline dropshot/src/websocket.rs:301]: If this is just here for the `impl JsonSchema for WebsocketUpgrade` we can probably just inline it.

File: dropshot/src/websocket.rs
Comment (ahl) [inline dropshot/src/websocket.rs:314]: is the schema really just to communicate this sentinel through?

File: dropshot/src/websocket.rs
Comment (ahl) [inline dropshot/src/websocket.rs:311]: I don't think this schema is ever really used in which case I think we could just replace this with
```suggestion
        let mut schema = schemars::schema::SchemaObject::default();
```

File: dropshot/src/api_description.rs
Comment (lifning) [reply dropshot/src/api_description.rs:1596]: i actually changed to this from a proper bare `impl Default` after an earlier PR comment pointed out that the rust-toolchain.toml for this project declares rust 1.62 - simple enough to revert to what i originally had if it's a concern at all

**Comment by lifning** [comment]
> Are other extractors permitted alongside `WebsocketUpgrade`? Or must it be the lone extractor?

They shouldn't be mutually exclusive (or else we wouldn't be able to do much useful with it, like say addressing a specific org/proj/inst in nexus by `Path<>` params)

> What does this look like in the OpenAPI output? The OpenAPI spec folks have pretty much declared Websockets as "not HTTP APIs" and therefore out of scope.

It looks like a GET endpoint with a `"x-dropshot-websocket": true` stashed in it, just like the `"x-dropshot-pagination": true` we already have. Here's an example (with the unrelated formatting condensed a little for brevity)
```json
{
  "/organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/serial-console": {
    "get": {
      "tags": ["instances"],
      "summary": "Connect to an instance's serial console stream",
      "operationId": "instance_serial_console",
      "parameters": [
        { "in": "path", "name": "instance_name", "required": true,
          "schema": { "$ref": "#/components/schemas/Name" }, "style": "simple" },
        { "in": "path", "name": "organization_name", "required": true,
          "schema": { "$ref": "#/components/schemas/Name" }, "style": "simple" },
        { "in": "path", "name": "project_name", "required": true,
          "schema": { "$ref": "#/components/schemas/Name" }, "style": "simple" }
      ],
      "responses": { "default": { "description": "", "content": { "*/*": {"schema": {}}}}},

      "x-dropshot-websocket": true
    }
  }
}
```

> I vaguely had intended to use [AsyncAPI](https://www.asyncapi.com/) to describe websocket interfaces, both the channels and the messages. AsyncAPI and OpenAPI are at least sort of compatible in that I think we could produce a single JSON file with both descriptions in there.

Would it be reasonable (in the interest of providing a websockets interface in the nearer term for serial & VNC API work) to punt on the full extent of that for the moment with the "x-dropshot-websocket" sentinel in an OpenAPI path's "get" method, and then follow up by implementing the proper AsyncAPI version?

For the use cases I'm thinking of, AsyncAPI's specifications are a bit heavier weight than we need, but that might be a blessing in this regard - I think we often want to forward *other* protocols through raw binary messages, so most of the work that I think this would end up unblocking would have a fairly trivial expression as far as just annotating everywhere it's used with an effective "yep, it's binary" in such a way that our macro could accept before having a "real" AsyncAPI implementation backing it.

> Does it make sense to use the same `endpoint` macro with the constraints that the method must be a `GET`, the `WebsocketUpgrade` must be present, and the return type is `Result<Response<Body>, HttpError>`? Or might we want some other interface such as `#[channel { path = ... } ]` or `#[endpoint { method = WEBSOCKET, ...`?

I did have the thought of perhaps making `endpoint { method = WEBSOCKET, ...` decorating an async fn in which you're required to put an upgraded connection parameter (much like requiring path variables), and not have an intermediate function at all... but I shyed away from the former part of that approach at first because I didn't want to use a non-HTTP-verb there, and now the latter part too because would mean we wouldn't have a good way to expose the JoinHandle of what will likely be a longer-running future than most other kinds of endpoint.

I also had the use case of "what if we want to leave a path open for hybrid GET endpoints requests that may or may not be upgraded depending on how it's accessed," but I am not sure this use case would be worth the additional complexity.

`#[channel...` would be a nice way to both get a little closer to AsyncAPI terminology and possibly isolate some of the additional websocket/AsyncAPI-specific complexity from the `endpoint` decorator's implementation when it comes to things like describing message formats. My current inclination in light of this is to perhaps make such a thing as a declarative macro for now, that expands to the currently-equivalent `#[endpoint = GET` decorator.

File: dropshot/src/api_description.rs
Comment (ahl) [reply dropshot/src/api_description.rs:1596]: I should have clarified. I don't know if we have an MSRV! Yes, I know we specify the toolchain; that's at least in part to ensure repeatable builds as we have several tests for the `endpoint` macros that are a bit.. finicky with regard to version dependence.

I love `#[default]`; I don't know if using it would be annoying for our internal consumers (or maybe for external consumers; 👋 I don't know if you exist, but if you do, we love you!)

**Comment by ahl** [comment]
> > Are other extractors permitted alongside `WebsocketUpgrade`? Or must it be the lone extractor?
> 
> They shouldn't be mutually exclusive (or else we wouldn't be able to do much useful with it, like say addressing a specific org/proj/inst in nexus by `Path<>` params)

Is it compatible with `Query` or `Body` extractors?

> > What does this look like in the OpenAPI output? The OpenAPI spec folks have pretty much declared Websockets as "not HTTP APIs" and therefore out of scope.
> 
> It looks like a GET endpoint with a `"x-dropshot-websocket": true` stashed in it, just like the `"x-dropshot-pagination": true` we already have. Here's an example (with the unrelated formatting condensed a little for brevity)

Cool!

> Would it be reasonable (in the interest of providing a websockets interface in the nearer term for serial & VNC API work) to punt on the full extent of that for the moment with the "x-dropshot-websocket" sentinel in an OpenAPI path's "get" method, and then follow up by implementing the proper AsyncAPI version?

Absolutely, and I apologize if I seemed to be implying otherwise. I don't want to stand in the way of urgent progress; I also want to have line of sight to how we want this to work in the future.

> For the use cases I'm thinking of, AsyncAPI's specifications are a bit heavier weight than we need, but that might be a blessing in this regard - I think we often want to forward _other_ protocols through raw binary messages, so most of the work that I think this would end up unblocking would have a fairly trivial expression as far as just annotating everywhere it's used with an effective "yep, it's binary" in such a way that our macro could accept before having a "real" AsyncAPI implementation backing it.

Neat. I haven't used AsyncAPI. I inferred that it lets you specify the protocol such that e.g. we could generate a client that had the messages as Rust types... but that's just imagineering.

> > Does it make sense to use the same `endpoint` macro with the constraints that the method must be a `GET`, the `WebsocketUpgrade` must be present, and the return type is `Result<Response<Body>, HttpError>`? Or might we want some other interface such as `#[channel { path = ... } ]` or `#[endpoint { method = WEBSOCKET, ...`?
> 
> I did have the thought of perhaps making `endpoint { method = WEBSOCKET, ...` decorating an async fn in which you're required to put an upgraded connection parameter (much like requiring path variables), and not have an intermediate function at all... but I shyed away from the former part of that approach at first because I didn't want to use a non-HTTP-verb there, and now the latter part too because would mean we wouldn't have a good way to expose the JoinHandle of what will likely be a longer-running future than most other kinds of endpoint.
> 
> I also had the use case of "what if we want to leave a path open for hybrid GET endpoints requests that may or may not be upgraded depending on how it's accessed," but I am increasingly unsure this use case would be worth the additional complexity.

Should we validate at least that folks are using a `GET` rather than some other method? I would be happy to help or to do this as a follow-on PR.

> `#[channel...` would be a nice way to both get a little closer to AsyncAPI terminology and possibly isolate some of the additional websocket/AsyncAPI-specific complexity from the `endpoint` decorator's implementation when it comes to things like describing message formats. My current inclination in light of this is to perhaps make such a thing as a trivial macro for now, that expands to the currently-equivalent `#[endpoint = GET` decorator.

That's a neat idea for a follow-on.

Thanks for the thorough discussion!

File: dropshot/src/api_description.rs
Comment (luqmana) [reply dropshot/src/api_description.rs:1596]: I did mention the derive Default but we don't have an MSRV I could find so depends on how conservative we wanna be here I guess.

In terms of our own internal consumers, of the ones I know of:
- propolis lists an [MSRV of 1.62](https://github.com/oxidecomputer/propolis/#minimum-supported-rust-version-msrv)
- omicron is pinned on an [old nightly](https://github.com/oxidecomputer/omicron/blob/9bbf2a0fa436542e6bfe2f403114ba280fe3730d/rust-toolchain.toml#L13) (due to unstable features): nightly-2022-04-27
- crucible is also pinned on an [even older nightly](https://github.com/oxidecomputer/crucible/blob/1eac1be81e0b3d624f2d09aecc40601a7aaef5f1/rust-toolchain.toml#L2): nightly-2021-11-24
- buildomat doesn't have a rust-toolchain or note about version
- cio is pinned on an [old nightly](https://github.com/oxidecomputer/cio/blob/209300a352cb653fac660c0234f80749b5974a02/rust-toolchain#L1): nightly-2022-03-13

Comment (ahl) [reply dropshot/src/api_description.rs:1596]: Do the old nightly's support #[default]? If so I say we go ahead and keep it. If not... I guess maybe roll them forward first?

Comment (lifning) [reply dropshot/src/api_description.rs:1596]: i think i'm not quite as inclined to have us drag everything forward for something as trivial as some very small syntax sugar (vs. say, implementing an actual *feature* in such a way that otherwise wouldn't have been possible), but i wouldn't  say no if other folks felt strongly otherwise

Comment (luqmana) [reply dropshot/src/api_description.rs:1596]: Yea, I don't think that that should hold this up.

**Comment by lifning** [comment]
so the little "panic in Drop if it's not handled" scheme is kinda invalidated when any of the other extractors fail to parse from the request - it never reaches the user's code for them to call `handle`, which would probably be pretty inconvenient even when the panic is downgraded to a `warn!` when built in release mode. i'm leaning toward just trying to writing the wrapper for it in the proc macro, and if the user needs a JoinHandle there's theoretically nothing stopping them from `tokio::spawn`ing inside the handler, though it would theoretically be a slight perf hit.

File: dropshot/src/websocket.rs
Comment (lifning) [reply dropshot/src/websocket.rs]: i think this is resolved by the new `#[channel...]` approach, as we don't let `method` be specified at all in that macro (it's always GET for protocol=WEBSOCKETS)

File: dropshot/src/websocket.rs
Comment (lifning) [reply dropshot/src/websocket.rs]: we don't explicitly check for it in proc-macro-land, but in the latest patch i did introduce some friendlier type aliases for what's going on - and the user need not concern themselves with the `Result<Response<Body>, HttpError>` one any more; that's all abstracted away by `#[channel]` and all their socket-handling function has to return is a very generic `Result<(), Box<dyn Error+Send+Sync+'static>` (chosen so we can log their errors, if any, and let them use the ergonomic `?`)

File: dropshot/src/api_description.rs
Comment (lifning) [reply dropshot/src/api_description.rs:1592]: per our discussion on matrix, i'll file some issues after this merges about turning this AsyncAPI support (and perhaps coming up with a general.. philosophy? on how to approach other extensions, like pagination or anything we come up with in the future that's outside of the scope of AsyncAPI in particular)

File: dropshot/src/handler.rs
Comment (ahl) [inline dropshot/src/handler.rs:206]: @lifning it looks like this was a breaking change:

```
error[E0560]: struct `dropshot::ExtractorMetadata` has no field named `paginated`
  --> nexus/src/authn/external/cookies.rs:51:29
   |
51 |         ExtractorMetadata { paginated: false, parameters: vec![] }
   |                             ^^^^^^^^^ `dropshot::ExtractorMetadata` does not have this field
   |
   = note: available fields are: `extension_mode`, `parameters`
```

https://github.com/oxidecomputer/omicron/pull/1626

Could you please update the changelog?

Comment (lifning) [reply dropshot/src/handler.rs:206]: https://github.com/oxidecomputer/dropshot/pull/428

**Pull request merged**
